crates/solidity/testing/sourcify/src/sourcify.rs

```diff
@@ -184,25 +205,37 @@ pub struct ContractArchive {
     /// Path to the directory inside this archive which contains all of the contracts.
     /// Iterate over the entries at this path to read the contracts.
     contracts_path: PathBuf,
+
+    /// Should remove the archive and contracts at the end
+    delete_files: bool,
 }
 
 impl ContractArchive {
-    pub fn fetch(desc: &ArchiveDescriptor) -> Result<ContractArchive> {
-        let client = Client::new();
-        let res = client.get(&desc.url).send()?;
-
-        let status = res.status();
-        if !status.is_success() {
-            bail!("Could not fetch source tarball");
+    pub fn fetch(desc: &ArchiveDescriptor, options: &ArchiveOptions) -> Result<ContractArchive> {
+        let contracts_path = desc.contracts_dir();
+        if !options.offline && !file_exists_and_is_fresh(&contracts_path) {
+            println!("Downloading shard archive from {url}", url = desc.url);
```

Comment:
rather than checking file age (24h), I wonder if we can use the `last-modified` or `etag` headers instead? `reqwest` already provides an example to do that, so we can skip downloading the manifest (or individual archives) if the contents didn't change. I suspect most of them are rarely updated within the 24 hours:

```rs
use reqwest::blocking::Client;
use reqwest::header::ETAG;

let client = Client::new();

let mut resp = client.get(URL).send()?;
if resp.status().is_success() {
    if existing_etag = resp.headers().get(ETAG_OR_LAST_UPDATED) {
        // no need to download
    } else {
        let mut file = std::fs::File::create("file")?;
        resp.copy_to(&mut file)?;
    }
}
```

Reply:
The intent was to avoid any network activity if the file was deemed fresh, but I think what you propose makes more sense. We'd still have `--offline` to avoid hitting sourcify's servers.